### ISSUE TITLE
Typescript support is missing (with workaround)

### ISSUE THREAD

**Issue by jclaveau** [blackflux/object-scan#1674]
Types definitions are missing for importing this module in a Typescript module.
```typescript
import objectScan from 'object-scan';
```
Triggers
```
Could not find a declaration file for module 'object-scan'. '<...>/node_modules/object-scan/lib/index.js' implicitly has an 'any' type.
  Try `npm i --save-dev @types/object-scan` if it exists or add a new declaration (.d.ts) file containing `declare module 'object-scan';`
```
Types definition package is missing:
```shell
yarn add @types/object-scan
yarn add v1.22.17
[1/6] Validating package.json...
[2/6] Resolving packages...
error An unexpected error occurred: "https://registry.yarnpkg.com/@types%2fobject-scan: Not found".
```

### Workaround
As a `workaround`, we can add a *.d.ts file anywhere our `/src` (or whatever source folder included in our `ts-config`). Personally I created a `/src/missing_types.d.ts` file containing 
```typescript
declare module 'object-scan';
```
But we then have to add a lot of `// @ts-ignore` for implicit 'any' types errors :
```
var property: any
Binding element 'property' implicitly has an 'any' type.
```

Example:
```typescript
      objectScan(['**'], {
        //@ ts-ignore
        filterFn: ({ property }) => {
          console.log('property', property);
        }
      })(myVar);
```
### Better Workaround
To avoid this and be a little cleaner, I wrote a basic type def file at `<reo>/src/@types/object-scan.d.ts` containing
```typescript
declare module 'object-scan';

declare namespace ObjectScan {

  export type PatternCollection = Array<String>
  export type Node = Array<any> | Object
  export type Key = Array<Number>
  export type Value = any

  // Following the order of https://github.com/blackflux/object-scan#options
  export interface ScannedItem {
    key: Key,
    value: Value,
    entry: [Key, Value],
    property: string,
    gproperty: string,

    parent: Node,
    gparent?: Node,
    parents: Array<Node>,

    isMatch: boolean,
    matchedBy: PatternCollection,
    excludedBy: PatternCollection,

    traversedBy: PatternCollection,
    isCircular: boolean,
    isLeaf: boolean,

    depth: Number,
    result: any,

    getKey: () => Key
    getValue: () => Value
    getEntry: () => [Key, Value]
    getProperty: () => string
    getGproperty: () => string

    getParent: () => Node
    getGparent: () => Node
    getParents: () => Array<Node>

    getIsMatch: () => boolean
    getMatchedBy: () => PatternCollection
    getExcludedBy: () => PatternCollection
    getTraversedBy: () => PatternCollection
    getIsCircular: () => boolean
    getIsLeaf: () => boolean
    getDepth: () => Number

    getResult: () => any

    context?: any,
  }
}
```
which I use the following way:
```typescript
export default class Something {
  test() {
    console.log('test', JSON.stringify(this))
  }
}
```

```typescript
import objectScan from 'object-scan';
// ...
      objectScan(['**'], {
        filterFn: (function({ key, property, value, parent }: ObjectScan.ScannedItem & {parent: Something}) {
          if (parent instanceof Something) {
            parent.test();
          }
        })
      })(stories);
```
I let some as `any` like `context` or `result` because I don't know how they work for now. Also I probably made some mistakes that I will fix when I'll use them.

### Long term solution
I never wrote a Typescript lib so I don't know if types must be inside or in a separate `@types/object-scan` installable but they seem to be some tools for that:
- https://github.com/Microsoft/dts-gen is available to generate automatically type defs:
- More at https://stackoverflow.com/questions/12687779/how-do-you-produce-a-d-ts-typings-definition-file-from-an-existing-javascript

Anyway, thank you for your work!

**Comment by simlu** [comment]
@jclaveau Hello and thank you for taking the time to write this!

We've made great progress on moving all blackflux packages from commonjs to ESM. Once that is done, typescript support shouldn't be that hard to add.

Typescript support is definitely an important goal and we are moving towards it!

**Comment by cdedreuille** [comment]
Any update on typescript support?

**Comment by simlu** [comment]
@cdedreuille Hello and thanks for asking! I was actually just looking into typescript last night now that this is the last enhancement in this project!

I have converted object-treeify to typescript (since it is a much simpler package) and the test are running and the build step is working (still dealing with @vercel/ncc, but that should be simple)

I'm still figuring out how to convert eslint and some other structure tests. I'm also new to typescript so will need to get a good review going once I have a pr up.

Once I understand all the configuration and setup and have that working it will be easy to convert this project as well. 

Tldr; So yes, expect this sooner rather than later!

**Comment by simlu** [comment]
Would be great if someone could take a look: https://github.com/blackflux/object-treeify/pull/1072

I'm kind of stuck, I'm getting: `Parsing error: "parserOptions.project" has been set for @typescript-eslint/parser`

While that error in theory should be easy to resolve - it somehow isn't.

**Comment by milahu** [comment]
> last enhancement in this project

full conversion to typescript is overkill
just add a *.d.ts file so the interface is typed

**Comment by christopher-caldwell** [comment]
Here's what I've used to make TS work. The above types didn't really help me that much.

```ts
/* eslint-disable @typescript-eslint/no-explicit-any */
declare module "object-scan" {
  type Needles = string[];

  interface CallbackArgs {
    /** key that callback is invoked for (respects joined option). */
    key: string | number;
    /** value for key. */
    value: any;
    /** entry consisting of [key, value]. */
    entry: any;
    /** current parent property. */
    property: any;
    /** current grandparent property. */
    gproperty: any;
    /** current parent. */
    parent: any;
    /** current grandparent. */
    gparent: any;
    /** array of form [parent, grandparent, ...]. */
    parents: any;
    /** true iff last targeting needle exists and is non-excluding. */
    isMatch: any;
    /** all non-excluding needles targeting key. */
    matchedBy: any;
    /** all excluding needles targeting key. */
    excludedBy: any;
    /** all needles involved in traversing key. */
    traversedBy: any;
    /** true iff value contained in parents */
    isCircular: any;
    /** true iff value can not be traversed */
    isLeaf: any;
    /** length of key */
    depth: any;
    /** intermittent result as defined by rtn */
    result: any;
    /** joined?: boolean): function that returns key */
    getKey: any;
    /** function that returns value */
    getValue: any;
    /** joined?: boolean): function that returns entry */
    getEntry: any;
    /** function that returns property */
    getProperty: any;
    /** function that returns gproperty */
    getGproperty: any;
    /** function that returns parent */
    getParent: any;
    /** function that returns gparent */
    getGparent: any;
    /** function that returns parents */
    getParents: any;
    /** function that returns isMatch */
    getIsMatch: any;
    /** function that returns matchedBy */
    getMatchedBy: any;
    /** function that returns excludedBy */
    getExcludedBy: any;
    /** function that returns traversedBy */
    getTraversedBy: any;
    /** function that returns isCircular */
    getIsCircular: any;
    /** function that returns isLeaf */
    getIsLeaf: any;
    /** function that returns depth */
    getDepth: any;
    /** function that returns result */
    getResult: any;
    /** as passed into the search */
    context: any;
  }

  type OptionFunction = (callbackArgs: CallbackArgs) => boolean;
  interface Options {
    joined: boolean;
    rtn: string | string[] | OptionFunction;
    reverse: boolean;
    orderByNeedles: boolean;
    abort: boolean;
    joined: boolean;
    useArraySelector: boolean;
    strict: boolean;
    filterFn: OptionFunction;
    breakFn: OptionFunction;
    beforeFn: OptionFunction;
    afterFn: OptionFunction;
    compareFn: OptionFunction;
  }

  type ObjectScan = <TInput, TReturn = unknown>(
    needles: Needles,
    options: Partial<Options>
  ) => (input: TInput) => TReturn;

  const objectScan: ObjectScan;

  export default objectScan;
}

```
